File: core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/twofactor/FetchDeliveryMethods.java
package org.mifos.mobilewallet.core.domain.usecase.twofactor;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository;
import org.mifos.mobilewallet.core.domain.model.twofactor.DeliveryMethod;

import java.util.List;

import javax.inject.Inject;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by ankur on 01/June/2018
 */

public class FetchDeliveryMethods extends
        UseCase<FetchDeliveryMethods.RequestValues, FetchDeliveryMethods.ResponseValue> {

    private final FineractRepository mFineractRepository;

    @Inject
    public FetchDeliveryMethods(FineractRepository fineractRepository) {
        mFineractRepository = fineractRepository;
    }

    @Override
    protected void executeUseCase(RequestValues requestValues) {
        mFineractRepository.getDeliveryMethods()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<List<DeliveryMethod>>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        getUseCaseCallback().onError(e.toString());
                    }

                    @Override
                    public void onNext(List<DeliveryMethod> deliveryMethods) {
                        getUseCaseCallback().onSuccess(new ResponseValue(deliveryMethods));
                    }
                });
    }

    public static final class RequestValues implements UseCase.RequestValues {

        public RequestValues() {

        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {

        private final List<DeliveryMethod> deliveryMethodList;

        public ResponseValue(
                List<DeliveryMethod> deliveryMethodList) {
            this.deliveryMethodList = deliveryMethodList;
        }

        public List<DeliveryMethod> getDeliveryMethodList() {
            return deliveryMethodList;
        }
    }
}

File: core/data/src/main/java/org/mifos/mobilewallet/core/domain/usecase/twofactor/FetchDeliveryMethods.kt
package org.mifos.mobilewallet.core.domain.usecase.twofactor

import org.mifos.mobilewallet.core.base.UseCase
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository
import org.mifos.mobilewallet.core.domain.model.twofactor.DeliveryMethod
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by ankur on 01/June/2018
 */
class FetchDeliveryMethods @Inject constructor(private val mFineractRepository: FineractRepository) :
    UseCase<FetchDeliveryMethods.RequestValues, FetchDeliveryMethods.ResponseValue>() {

    class RequestValues : UseCase.RequestValues
    class ResponseValue(
        val deliveryMethodList: List<DeliveryMethod?>
    ) : UseCase.ResponseValue

    override fun executeUseCase(requestValues: RequestValues) {
        mFineractRepository.deliveryMethods
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<List<DeliveryMethod?>>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    useCaseCallback.onError(e.toString())
                }

                override fun onNext(deliveryMethods: List<DeliveryMethod?>) {
                    useCaseCallback.onSuccess(ResponseValue(deliveryMethods))
                }

            })
    }
}
